fsrs/optimizer.py

```diff
@@ -386,6 +383,303 @@ def _update_parameters(
+            avg_easy_review_duration = probs_and_costs_dict["avg_easy_review_duration"]
+
+            simulation_cost = 0
+            for i in range(num_cards_simulate):
+                card = Card()
+                curr_date = start_date
+                while curr_date < end_date:
+                    # the card is new
+                    if curr_date == start_date:
+                        rating = rng.choices(
+                            [Rating.Again, Rating.Hard, Rating.Good, Rating.Easy],
+                            weights=[
+                                prob_first_again,
+                                prob_first_hard,
+                                prob_first_good,
+                                prob_first_easy,
+                            ],
+                        )[0]
+
+                        if rating == Rating.Again:
+                            simulation_cost += avg_first_again_review_duration
+
+                        elif rating == Rating.Hard:
+                            simulation_cost += avg_first_hard_review_duration
+
+                        elif rating == Rating.Good:
+                            simulation_cost += avg_first_good_review_duration
+
+                        elif rating == Rating.Easy:
+                            simulation_cost += avg_first_easy_review_duration
+
+                    # the card is not new
+                    else:
+                        rating = rng.choices(
+                            ["recall", Rating.Again],
+                            weights=[desired_retention, 1.0 - desired_retention],
+                        )[0]
+
+                        if rating == "recall":
+                            # compute probability that the user chose hard/good/easy, GIVEN that they correctly recalled the card
+                            rating = rng.choices(
+                                [Rating.Hard, Rating.Good, Rating.Easy],
+                                weights=[prob_hard, prob_good, prob_easy],
+                            )[0]
+
+                        if rating == Rating.Again:
+                            simulation_cost += avg_again_review_duration
+
+                        elif rating == Rating.Hard:
+                            simulation_cost += avg_hard_review_duration
+
+                        elif rating == Rating.Good:
+                            simulation_cost += avg_good_review_duration
+
+                        elif rating == Rating.Easy:
+                            simulation_cost += avg_easy_review_duration
+
+                    card, _ = scheduler.review_card(
+                        card=card, rating=rating, review_datetime=curr_date
+                    )
+                    curr_date = card.due
+
+            total_knowledge = desired_retention * num_cards_simulate
+            simulation_cost = simulation_cost / total_knowledge
+
+            return simulation_cost
```

Comment:
`simulation_cost = simulation_cost / total_knowledge` is confusing.

Reply:
To me it seems pretty clear. Here, we're computing the cost of 1000 simulated cards and at the end, we normalize the cost by total knowledge.

Do you have better alternative in mind?

Reply:
Fine. Let's merge it.